frontend/src/metabase/common/components/CollectionPicker/hooks.ts

```diff
@@ -0,0 +1,166 @@
+  location: "/",
+  description: "",
+  here: ["collection"],
+  below: ["collection"],
+};
+
+/**
+ * This is a special item list that exists "above" our analytics and might include:
+ * a) the highest-level collections the user can access (often "our analytics")
+ * b) the user's personal collection
+ * c) a top level folder including all personal collections (admin only)
+ */
+export const useRootCollectionPickerItems = (
+  options: CollectionItemListProps["options"],
+) => {
+  const isAdmin = useSelector(getUserIsAdmin);
+  const currentUser = useSelector(getUser);
+
+  const { data: personalCollection, isLoading: isLoadingPersonalCollecton } =
+    useGetCollectionQuery(
+      currentUser?.personal_collection_id
+        ? { id: currentUser.personal_collection_id }
+        : skipToken,
+    );
+
+  const {
+    data: personalCollectionItems,
+    isLoading: isLoadingPersonalCollectionItems,
+  } = useListCollectionItemsQuery(
+    currentUser?.personal_collection_id
+      ? {
+          id: currentUser?.personal_collection_id,
+          models: ["collection"],
+        }
+      : skipToken,
+  );
+
+  const {
+    data: rootCollection,
+    isLoading: isLoadingRootCollecton,
+    error: rootCollectionError,
+  } = useGetCollectionQuery({ id: "root" });
+
+  const items = useMemo(() => {
+    const collectionItems: CollectionPickerItem[] = [];
+
+    if (options.showRootCollection || options.namespace === "snippets") {
+      if (rootCollection && !rootCollectionError) {
+        collectionItems.push({
+          ...rootCollection,
+          model: "collection",
+          here: ["collection"],
+          location: "/",
+          name:
+            options.namespace === "snippets"
+              ? t`Top folder`
+              : rootCollection.name,
+        });
+      } else if (rootCollectionError) {
+        collectionItems.push({
+          name: t`Collections`,
+          id: "root",
+          here: ["collection"],
+          description: null,
+          can_write: false,
+          model: "collection",
+          location: "/",
+        });
+      }
+    }
+
+    if (
+      options.showPersonalCollections &&
+      options.namespace !== "snippets" &&
+      currentUser &&
+      !!personalCollection
+    ) {
+      collectionItems.push({
+        ...personalCollection,
+        here: personalCollectionItems?.data.length ? ["collection"] : [],
+        model: "collection",
+        can_write: true,
+      });
+
+      if (isAdmin) {
+        collectionItems.push(personalCollectionsRoot);
+      }
+    }
+
+    return collectionItems;
+  }, [
+    currentUser,
+    personalCollection,
+    rootCollection,
+    isAdmin,
+    options,
+    rootCollectionError,
+    personalCollectionItems,
+  ]);
+
+  const isLoading =
+    isLoadingRootCollecton ||
+    isLoadingPersonalCollecton ||
+    isLoadingPersonalCollectionItems;
+
+  return { items, isLoading };
+};
+
+export const useEnsureCollectionSelected = ({
```

Comment:
I'm not happy about this hook, it's quite complex.

Reply:
You can extract this part to a hook
```
const {
    data: rootCollection,
    isLoading: isLoadingRootCollecton,
    error: rootCollectionError,
  } = useGetCollectionQuery({ id: "root" });

  const items = useMemo(() => {
    const collectionItems: CollectionPickerItem[] = [];

    if (options.showRootCollection || options.namespace === "snippets") {
      if (rootCollection && !rootCollectionError) {
        collectionItems.push({
          ...rootCollection,
          model: "collection",
          here: ["collection"],
          location: "/",
          name:
            options.namespace === "snippets"
              ? t`Top folder`
              : rootCollection.name,
        });
      } else if (rootCollectionError) {
        collectionItems.push({
          name: t`Collections`,
          id: "root",
          here: ["collection"],
          description: null,
          can_write: false,
          model: "collection",
          location: "/",
        });
      }
    }
```